src/ess/bifrost/io/sqw.py

```diff
@@ -0,0 +1,424 @@
+    """Histogram the events in (detector, setting, incident_energy).
+
+    This also converts the weights to float32 as no further reductions
+    will be performed for the 'pixel data'.
+    We accept a loss of precision for the 'image data' here which will be
+    converted back to float64 later.
+    """
+    hist = flat_events.hist(incident_energy=energy_bins)
+    hist.coords['incident_energy'] = sc.midpoints(hist.coords['incident_energy']).to(
+        dtype='float32', copy=False
+    )
+    hist.coords['ien'] = sc.arange(
+        'incident_energy',
+        1,  # 1-based indexing
+        hist.sizes['incident_energy'] + 1,
+        dtype='float32',
+        unit=None,
+    )
+
+    hist.data = hist.data.to(dtype='float32', copy=False)
+
+    return hist
+
+
+def _with_inelastic_coords(
+    observations: sc.DataArray, gravity: GravityVector
+) -> sc.DataArray:
+    """Compute and assign Qx, Qy, Qz, and energy_transfer.
+
+    This also drops all coordinates that are no longer needed.
+    """
+    from ess.spectroscopy.indirect.conversion import (
+        energy_transfer,
+        lab_momentum_transfer_from_incident_energy,
+        rotate_to_sample_table_momentum_transfer,
+    )
+
+    graph = {
+        'lab_momentum_transfer': lab_momentum_transfer_from_incident_energy,
+        'sample_table_momentum_transfer': rotate_to_sample_table_momentum_transfer,
+        'energy_transfer': energy_transfer,
+        'gravity': lambda: gravity,
+    }
+    aux = observations.transform_coords(
+        ['sample_table_momentum_transfer', 'energy_transfer'],
+        graph=graph,
+        keep_inputs=False,
+        keep_intermediate=False,
+        quiet=True,
+    )
+    q = aux.coords.pop('sample_table_momentum_transfer')
+
+    return aux.assign_coords(
+        energy_transfer=aux.coords.pop('energy_transfer').to(
+            dtype='float32', copy=False
+        ),
+        **{
+            f'Q{dim}': _project_onto(q, axis).to(dtype='float32')
+            for dim, axis in zip('xyz', (_AXIS_U, _AXIS_V, _AXIS_W), strict=True)
+        },
+    )
+
+
+def _project_onto(vec: sc.Variable, axis: sc.Variable) -> sc.Variable:
+    axis = axis / sc.norm(axis)
+    # Optimization to avoid huge dot-products:
+    match list(axis.values):
+        case [1, 0, 0]:
+            return vec.fields.x
+        case [0, 1, 0]:
+            return vec.fields.y
+        case [0, 0, 1]:
+            return vec.fields.z
+        case _:
+            return sc.dot(axis, vec)
+
+
+def _bin_image(
+    observations: sc.DataArray, bin_sizes: SQWBinSizes
+) -> tuple[sc.DataArray, sc.Variable, sc.DataArray]:
+    # Dim order to match pixel data (note the transpose for dnd)
+    binned = observations.bin(
+        {dim: bin_sizes[dim] for dim in ('energy_transfer', 'Qz', 'Qy', 'Qx')}
+    ).rename(energy_transfer='u4', Qz='u3', Qy='u2', Qx='u1')
+
+    image = binned.transpose(['u1', 'u2', 'u3', 'u4'])
+    dnd = image.bins.mean().to(dtype='float64', copy=False)
```

Comment:
Is this the place we do a `mean` instead of a `sum` to match the Horace convention?
If so, it might be worth adding a comment about this?

Reply:
Yes, will do.